### ISSUE TITLE
Dose the data division for multi-gpu destruct the accuracy?

### ISSUE THREAD

**Issue by venuswu** [tensorflow/tensorflow#33369]
Data division for multi-gpu after tf1.11 seems to have the same shape of data across all gpus, which will result in a bigger variance of loss between different steps. For example, for the transformer model, when in one step, all the gpus are feeded with the data with length near 8, but in the next step, all the gpus are feeded with the data with length near 32, which will result in a bigger variance of loss.

In practice, I got a lower accuracy trained with tf2.0 than tf1.11.

@guptapriya @yuefengz @zongweiz @sgpyc

**Comment by guptapriya** [comment]
Hi,

Thank you for your question. Can you elaborate on your example? Why is the
data length 8 vs 32 in subsequent steps? Do you mean you have dynamic input
per step?

We have been able to train transformer to the state of art accuracy with
2.0 and 1.15, but we havent tested with 1.11. can you share your setup?
Which API are you you using, what are the batch sizes etc?

**Comment by venuswu** [comment]
Yes, I used dynamic batchsize to train the official/transformer.
https://github.com/tensorflow/models/blob/master/official/transformer/v2/data_pipeline.py

Here is the data pipeline of 2.0 same as 1.11, for which batchsize means the number of tokens.  For example in tf1.11, I set the batchsize to 4096 which means 4096 tokens for every replica, for a replica       the length of the tokens seem to be as near as possible(padding fills up the gap), for example a replica can get a batch with the shape of [1024,4], and another get a shape of [ 128,32], in the same step, the shape of the batch may vary between different replica. But there is similar situation between steps.
For the later version like tf1.14 and tf2.0, the parameter batchsize means total tokens of all the replicas, so for example the batchsize 4096*8 may have a shape of [8192,4] and autoshard across all the replicas, every replica get the same shape of [1024,4], but for the next step, it may read a shape of [1024, 32], and every replica get the same shape of [128,32], which may result in a variance between different steps.

I have read a paper https://arxiv.org/pdf/1806.00187.pdf
![image](https://user-images.githubusercontent.com/3830256/66908721-27548580-f03e-11e9-85fc-e4ffb4eced36.png)
 in which it is said that it my destruct the accuracy with all replica feeded with the same shape of data.

I don't kown whether I have described the question clearly.
@guptapriya @ravikyram

**Comment by venuswu** [comment]
`python transformer_main.py --data_dir=$DATA_DIR --model_dir=$MODEL_DIR \
    --vocab_file=$VOCAB_FILE --param_set=$PARAM_SET \
    --train_steps=$TRAIN_STEP --steps_between_evals=$TRAIN_STEP \
    --batch_size=4096 --max_length=256 \
    --num_gpus=8 \
    --epochs=275 \
    --enable_time_history=True \
    --enable_xla=False \
    --enable_grappler_layout_optimizer=True
`
Here is the command line. 
@guptapriya @ravikyram

**Comment by venuswu** [comment]
Is there a option for Backward compatibility? @guptapriya @ravikyram

**Comment by gowthamkpr** [comment]
@venuswu Is there a way that you can provide us a minimal reproducible example ? Also can you please explain it clearly as I am unable to understand what you are trying to convey? Whats the root cause of the issue?

**Comment by gowthamkpr** [comment]
Clsoing this issue as it has been inactive for more than 14 days. Please add additional details and we can open this issue again
